feat(lsp): code lenses

matoous opened the pull request:
Initial (limited) implementation of *LSP code lenses*[^1]. This is purposefully as small of a change as possible, it displays code lenses in the gutter and adds a single new command `show_code_lenses_under_cursor` that displays a picker for code lenses on the current line. Notably, displaying code lenses as annotations for the current line is not implemented and this implementation doesn't support resolving code lenses from multiple language servers. I am happy to address both but would prefer to do so separately to make this PR reviewable amidst the low bandwith and large backlog for maintainers.

## Code Lens

From the VS Code features docs[^vs_code]:

> The essence of the feature is "actionable contextual information interspersed" in your source code. That's quite a mouthful. Let me break it down for you. `CodeLens` are links in your code:
> Actionable - You can click on the link and something happens.
> Contextual - The links are close to the code they are representing.
> Interspersed - The links located throughout your source.

And from the official LSP docs[^lsp_req]:

>  A code lens represents a command that should be shown along with
> source text, like the number of references, a way to run tests, etc.

## Support

The support for code lens varies among LSPs but there are already nice features that can be achieved via code lenses (some of them listed below). The biggest limitation is that some language servers rely on client-side command execution which is usually solved via plugins. For other LSPs code lenses are opt-in only and without plugins can for example only display inline hints about number of references or number of implementations[^typescript].

Still, LSP is part of Helix and given Code Lenses are part of the specs, I think it's fair to implement them. Even if - without a plugin system - with limited capabilities.

- [golang](https://go.googlesource.com/tools/+/refs/heads/master/gopls/doc/settings.md#codelenses-map_string_bool)
  - `generate`, `vendor`, `update dependency`, and more.
  - test execution isn't supported[^2]
- [rust-analyzer](https://rust-analyzer.github.io/manual.html)
  - `run` for tests and main - requires client side execuity
  - `references`/`implementations` (show number of references/implementations and can be opened to view them, disabled by default) - requires client side execution
- [deno](https://deno.land/manual@v1.28.1/references/vscode_deno#code-lenses)
  - `run test` support - requires client side execution

(didn't check more languages).

## Future work

As mentioned above this implementation is limited and doesn't provide Code Lens as they should be provided - displayed contextually as a text - close to the code that they apply for. There should be a follow up PRs for:

- supporting multiple language servers
- displaying the code lenses as a line annotations

## Showcase

https://github.com/user-attachments/assets/e902a175-9dbf-4f42-9898-22f14d9b51b8

---

Related/fixes: https://github.com/helix-editor/helix/issues/3005
Related: https://github.com/helix-editor/helix/issues/3272

[^1]: https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_codeLens
[^2]: https://github.com/golang/go/issues/36787
[^vs_code]: https://code.visualstudio.com/blogs/2017/02/12/code-lens-roundup
[^lsp_req]: https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_codeLens
[^typescript]: https://github.com/typescript-language-server/typescript-language-server?tab=readme-ov-file#code-lenses-textdocumentcodelens

sigmaSd commented:
>rust-analyzer doesn't support executing commands (on the server side) so the commands need to be executed client side.

This is not specific  to rust-analyzer, I think that other lsps work the same way (especially for test codelens), for example I know its the case with deno lsp.

I think this would require plugins to handle, because executing the commands is lsp dependent

matoous commented:
@sigmaSd yes, indeed, you are right. Rust for example returns `rust-analyzer.showReferences` which is custom extension command which would indeed require plugins. I will push a few more changes but otherwise I guess this will need to be put on hold until the plugin system is available. Other option would be for the commands to be for now mapped to the shell via `languages.toml` configuration and the output being display the same way we do for `:run-shell-command`.

pascalkuthe commented:
`:sh` currently does not return anything if the command returns an error (which is the interesting case for tests I would assume) aslo any tests that might want input will freeze the editor. I think for running tests to be a good experience we would need an integrated terminal #4649.  Also going the `sh` route requires developing, maintaining and installing a separate CLI tool for each LSP command (these binaries don't exist yet) which I don't see people developing (and would only work for a subset of commands I guess). 

How many servers actually support server side execution? If rust-analyzer is the odd-one out with client-only commands then this might be worth merging but if almost all LSPs only allow client-side command execution then I think this should just probably wait for plugins.

ayoub-benali commented:
> > rust-analyzer doesn't support executing commands (on the server side) so the commands need to be executed client side.
> 
> This is not specific to rust-analyzer, I think that other lsps work the same way (especially for test codelens), for example I know its the case with deno lsp.
> 
> I think this would require plugins to handle, because executing the commands is lsp dependent

I can confirm that at least [Metals](https://scalameta.org/metals/) also expects the client to run some code lens commands and they are implemented in the plugin for sublime for example.

Esnos33 commented:
Hi, I'm new to git and I have question. I already have helix 23.03 and in release notes there is information about Initial support for snippets, but #3005 bug is still there. This is how it is supposed to be? Automatic type annotation is one of best things in haskell.

matoous commented:
Closing this until Helix has support for plugins that are prerequisit for this being useful.

maralorn commented:
I have a question about this: It is my impression that there are code lenses which don’t need a plugin to work. I think haskell-language-server has some of those. Is that wrong? Or could we maybe implement this, but only show code lenses in the case where they don’t require a plugin?
Actually, haskell-language-server simply offers code actions for when you click on the code lens. And those code actions can already be used right now. I am primarily interested in the code lenses because they display useful information, even if you don’t want to activate them right now.

alexaandru commented:
Similar to the above comment, in `gopls` there are code actions that are available after running code lenses and there already is support for code actions so we wouldn't need any plugins to use them.

For example, when in `go.mod` file, there is a code lense `Check for upgrades` which when run it will populate diagnostics with info about which packages can be upgraded. Then for each of those lines, there is a code action available to upgrade it. So code lenses alone would be a very valuable addition.

matoous commented:
👋 let me take a look if I could revive this. Indeed, the support might vary language to language but for some languages this might be worthwhile even in limited capacity.

maralorn commented:
Also: Even if Code lenses can’t be triggered it might still be useful to show them, right? And that part should always work without plugin support, right?

maralorn commented:
I am test driving this and it works just fine with haskell-language-server. Of course it would be nice if the code lenses would auto-update, but this is good enough for me.

Thank you!

matoous committed:
feat(lsp): code lens support

matoous committed:
feat(view): render lenses in gutter

matoous committed:
feat: request_code_lenses command

Allows requesting code lenses and renders them in the gutter
with the lowest priority.

matoous committed:
feat: code lenses handler

Request code lenses on newly opened documents,
when a language server is started, etc.

matoous committed:
chore: cleanup

matoous committed:
feat: resolve codelenses when needed

matoous committed:
fix: comments

matoous committed:
fix(docs): generate latest docs

matoous commented:
Heya 👋 I picked this up again after a loooong time, there seems to be way more of the ground work done to get this working properly. Rebase was too hard, so I started from scratch, sorry for the nuked history of this PR.

matoous committed:
feat(lsp): support workspace/codeLens/refresh server request

matoous committed:
chore: cleanup

matoous committed:
feat: update code lens positions on change

matoous committed:
fix: preserve code lenses on resolve
